File: src/c04_professional_classes/interface_segregation/refactored/InterfaceSegregationRefactored.java
package c04_professional_classes.interface_segregation.refactored;/*
In this example:

- We split the interfaces into Workable, Eatable, and Sleepable.
- HumanWorker implements all three interfaces, while RobotWorker implements only Workable.
- Each class now only depends on the methods it needs, adhering to the Interface Segregation Principle and resulting in a cleaner design.
*/

interface Workable {
    void work();
}

interface Eatable {
    void eat();
}

interface Sleepable {
    void sleep();
}

class HumanWorker implements Workable, Eatable, Sleepable {
    @Override
    public void work() {
        System.out.println("Human is working...");
    }

    @Override
    public void eat() {
        System.out.println("Human is eating...");
    }

    @Override
    public void sleep() {
        System.out.println("Human is sleeping...");
    }
}

class RobotWorker implements Workable {
    @Override
    public void work() {
        System.out.println("Robot is working...");
    }
    // RobotWorker doesn’t need to implement `eat` and `sleep` interfaces.
}

public class InterfaceSegregationRefactored {
    public static void main(String[] args) {
        HumanWorker human = new HumanWorker();
        RobotWorker robot = new RobotWorker();

        human.work();
        human.eat();
        human.sleep();

        robot.work();
    }
}

File: src/c04_professional_classes/matlab/InterfaceSegregationRefactored.m
function InterfaceSegregationRefactored()
    % In this example:
    % - We split the interfaces into Workable, Eatable, and Sleepable.
    % - HumanWorker implements all three, while RobotWorker only Workable.
    % - Each class only depends on what it needs, adhering to ISP.

    human = HumanWorkerRef();
    robot = RobotWorkerRef();

    human.work();
    human.eat();
    human.sleep();

    robot.work();
end

classdef IWorkable < handle
    methods (Abstract)
        work(obj)
    end
end

classdef IEatable < handle
    methods (Abstract)
        eat(obj)
    end
end

classdef ISleepable < handle
    methods (Abstract)
        sleep(obj)
    end
end

% MATLAB supports multiple inheritance
classdef HumanWorkerRef < IWorkable & IEatable & ISleepable
    methods
        function work(obj)
            fprintf('Human is working...\n');
        end
        function eat(obj)
            fprintf('Human is eating...\n');
        end
        function sleep(obj)
            fprintf('Human is sleeping...\n');
        end
    end
end

classdef RobotWorkerRef < IWorkable
    methods
        function work(obj)
            fprintf('Robot is working...\n');
        end
    end
end
